File: HisabCore/Sources/HisabCore/YearMonth.swift
import Foundation

/// A calendar month, resolved in Indian Standard Time.
public struct YearMonth: Hashable, Comparable, Codable, Sendable, CustomStringConvertible {
    public let year: Int
    public let month: Int  // 1...12

    public static var istCalendar: Calendar {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "Asia/Kolkata")!
        return cal
    }

    public init(year: Int, month: Int) {
        precondition((1...12).contains(month), "month out of range: \(month)")
        self.year = year
        self.month = month
    }

    public init(date: Date) {
        let comps = Self.istCalendar.dateComponents([.year, .month], from: date)
        self.init(year: comps.year!, month: comps.month!)
    }

    public func advanced(by months: Int) -> YearMonth {
        let total = year * 12 + (month - 1) + months
        let y = total >= 0 ? total / 12 : (total - 11) / 12
        return YearMonth(year: y, month: total - y * 12 + 1)
    }

    public static func months(from: YearMonth, through: YearMonth) -> [YearMonth] {
        guard from <= through else { return [] }
        var result: [YearMonth] = []
        var current = from
        while current <= through {
            result.append(current)
            current = current.advanced(by: 1)
        }
        return result
    }

    public static func < (lhs: YearMonth, rhs: YearMonth) -> Bool {
        (lhs.year, lhs.month) < (rhs.year, rhs.month)
    }

    public var description: String {
        String(format: "%04d-%02d", year, month)
    }

    private static let monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
                                     "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

    public var displayName: String {
        "\(Self.monthNames[month - 1]) \(year)"
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/year_month.dart
/// A calendar month, resolved in Indian Standard Time.
///
/// IST is a fixed UTC+05:30 offset with no DST, so all IST calendar math is
/// done by shifting UTC instants — no timezone database needed.
library;

const Duration istOffset = Duration(hours: 5, minutes: 30);

/// The instant shifted so its UTC fields read as IST wall-clock fields.
DateTime istClock(DateTime date) => date.toUtc().add(istOffset);

/// "yyyy-MM-dd" of the instant in IST — the day key used in content hashes
/// and synthetic references.
String istDayString(DateTime date) {
  final c = istClock(date);
  return '${c.year.toString().padLeft(4, '0')}-'
      '${c.month.toString().padLeft(2, '0')}-'
      '${c.day.toString().padLeft(2, '0')}';
}

/// "yyyyMMdd" of the instant in IST (synthetic reference recipe).
String istCompactDayString(DateTime date) =>
    istDayString(date).replaceAll('-', '');

class YearMonth implements Comparable<YearMonth> {
  final int year;
  final int month; // 1...12

  YearMonth(this.year, this.month) {
    if (month < 1 || month > 12) {
      throw ArgumentError('month out of range: $month');
    }
  }

  factory YearMonth.fromDate(DateTime date) {
    final c = istClock(date);
    return YearMonth(c.year, c.month);
  }

  YearMonth advancedBy(int months) {
    final total = year * 12 + (month - 1) + months;
    final y = total >= 0 ? total ~/ 12 : (total - 11) ~/ 12;
    return YearMonth(y, total - y * 12 + 1);
  }

  static List<YearMonth> monthsFromThrough(YearMonth from, YearMonth through) {
    if (from.compareTo(through) > 0) return [];
    final result = <YearMonth>[];
    var current = from;
    while (current.compareTo(through) <= 0) {
      result.add(current);
      current = current.advancedBy(1);
    }
    return result;
  }

  @override
  int compareTo(YearMonth other) =>
      (year * 12 + month) - (other.year * 12 + other.month);

  bool operator <(YearMonth other) => compareTo(other) < 0;
  bool operator <=(YearMonth other) => compareTo(other) <= 0;
  bool operator >(YearMonth other) => compareTo(other) > 0;
  bool operator >=(YearMonth other) => compareTo(other) >= 0;

  @override
  bool operator ==(Object other) =>
      other is YearMonth && other.year == year && other.month == month;

  @override
  int get hashCode => Object.hash(year, month);

  @override
  String toString() =>
      '${year.toString().padLeft(4, '0')}-${month.toString().padLeft(2, '0')}';

  static const _monthNames = [
    'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
    'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
  ];

  String get displayName => '${_monthNames[month - 1]} $year';
}
